File: src/leetCode/backTrack/Problem078.java
package leetCode.backTrack;

import java.util.ArrayList;
import java.util.List;

/**
 * Given a set of distinct integers, nums, return all possible subsets (the power set).
 * <p>
 * Note: The solution set must not contain duplicate subsets.
 * <p>
 * Example:
 * <p>
 * Input: nums = [1,2,3]
 * Output:
 * [
 * [3],
 * [1],
 * [2],
 * [1,2,3],
 * [1,3],
 * [2,3],
 * [1,2],
 * []
 * ]
 * create by stephen on 2018/6/26
 */
public class Problem078 {

    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        backTrack(result, new ArrayList<>(), nums, 0);
        return result;
    }

    private void backTrack(List<List<Integer>> result, List<Integer> tempList, int[] nums, int start) {
        result.add(new ArrayList<>(tempList));
        for (int i=start; i<nums.length; ++i) {
            tempList.add(nums[i]);
            backTrack(result, tempList, nums, i +1);
            tempList.remove(tempList.size() -1);
        }
    }


}

File: leetCode/backTrack/Problem078.cpp

#include <vector>

/**
 * Given an integer array nums of unique elements, return all possible subsets
 * (the power set).
 *
 * The solution set must not contain duplicate subsets. Return the solution in
 * any order.
 *
 */
void backtrack(std::vector<std::vector<int>>&, std::vector<int>&, int,
               std::vector<int>&);
std::vector<std::vector<int>> subsets(std::vector<int>& nums) {
  std::vector<std::vector<int>> result;
  std::vector<int> current;
  backtrack(result, nums, 0, current);
  return result;
}

void backtrack(std::vector<std::vector<int>>& result, std::vector<int>& nums,
               int idx, std::vector<int>& current) {
  result.push_back(current);
  for (int i = idx; i < nums.size(); ++i) {
    current.push_back(nums[i]);
    backtrack(result, nums, i + 1, current);
    current.pop_back();
  }
}
